src/ledger/LedgerStateSnapshot.h

```diff
@@ -0,0 +1,179 @@
+// Copyright 2024 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#pragma once
+
+#include "bucket/BucketListSnapshot.h"
+#include "bucket/BucketManager.h"
+#include "bucket/BucketSnapshotManager.h"
+#include "ledger/LedgerTxn.h"
+#include "util/NonCopyable.h"
+
+namespace stellar
+{
+
+class Application;
+class TransactionFrame;
+
+// A unified ledger entry interface that supports LedgerEntry representations
+// for both legacy SQL and BucketList snapshots. When working with LedgerTxn,
+// using plain LedgerEntry is not safe, as it might be modified or invalidated
+// by a nested transaction. To address this, the LedgerTxnEntry and
+// ConstLedgerTxnEntry abstractions should be preserved, allowing bucket
+// snapshot LedgerEntries to support the same interface.
+class LedgerEntryWrapper
+{
+    // Either hold a reference or a pointer to the entry
+    std::variant<LedgerTxnEntry, ConstLedgerTxnEntry,
+                 std::shared_ptr<LedgerEntry>>
+        mEntry;
+
+  public:
+    LedgerEntryWrapper(ConstLedgerTxnEntry&& entry);
+    LedgerEntryWrapper(LedgerTxnEntry&& entry);
+    LedgerEntryWrapper(std::shared_ptr<LedgerEntry> entry);
+    LedgerEntry& currentToModify();
```

Comment:
It looks like this is only used in loadgen, so it can be behind a `BUILD_TESTS` ifdef.

Reply:
Ended up not needing this function at all :)